bn-apps/cordapp-updates-distribution/design/design.md

```diff
@@ -0,0 +1,149 @@
+
+> CDS won't natively support *release channels*. Release channels can be simulated by using multiple Maven repositories (a repo per channel). Users can sync down different remote repositories to different local locations and then manually install a desired version of CorDapp to their nodes.
+
+##### cds-lib
+
+`cds-lib` is a convenience wrapper around Maven Resolver. It will support full Maven coordinates, in the format of `<groupId>:<artifactId>[:<extension>[:<classifier>]]:<version>` as well as version ranges. Version ranges will be specifiable in mathematical range notation, i.e. "[1.0,2.0)", "[1.0,)" or "[1.0]". Range queries are supported by Maven Resolver out-of-the-box and will not require extra development efforts. `cds-lib` will allow to:
+* Fetch metadata about a single or a range of versions from a remote Maven repository.
+* Download a single or a range of versions from a remote repository. `cds-lib` is packaging-type agnostic and can be used to pull down any type of artifacts from any Maven2 compatible repository. Support of classifiers will allow devs to publish CorDapps targeting different hardware / software configurations.
+
+`cds-lib` will support pluggable transports:
+* *HTTP(s)*. Available in Maven Resolver out-of-the-box with proxy- and repository- level authentications support.
+* *Corda Flows*. Transport implementation that allows transferring files over Corda Flows.
+* *Corda RPC*. Its essentially the same as *Corda Flows*, with the difference that the *RPC* version will be used if `cds-lib` is invoked from outside of a Corda node.
+
+`cds-lib` will be invokable from a command line, with command line parameters taking precedence over the configuration file. CLI invocation *could* look like:
+
+```
+# Will pull down a single version of cordapp and its dependencies
+> java -jar cds-lib.jar
+    -Dcordapp="com.my.company.name:corda-name:1.0"
+    -DremoteRepo="https://my-company-repo.com"
+    -DlocalRepo="/path/to/my/local/repo"
+    -Dtransport="rpc"
+    -Dconfiguration="config.properties"
+
+# Will pull down all versions of cordapps and their dependencies
+> java -jar cds-lib.jar
+    -Dcordapp="com.my.company.name:corda-name:[0,)"
+    -DremoteRepo="https://my-company-repo.com"
+    -DlocalRepo="/path/to/my/local/repo"
+    -Dtransport="http"
+    -Dconfiguration="config.properties"
+
+```
+
+Exact configuration parameters will be documented during the implementation.
+
+##### cds-cordapp
+
+`cds-cordapp` will provide the following functionality on *top* of the `cds-lib`:
+* Scheduled state for members to periodically sync their local repos with a single remote repository.
+* Flows for BNO to notify members about new CorDapp version availability. Members CorDapp will automatically attempt to download a new version.
+* Flows for BNO to notify members about CorDapp version revocations. If a version of CorDapp was revoked, members are expected to manually update their nodes to the last not-revoked version ASAP.
```

Comment:
What does the flow really do then, if the only outcome is manual? Maybe for now we can descope revocations and just ask them to email their users and tell them to downgrade?

Reply:
You mean descope revocations at all? Or just the "notification" part of it?
I agree re: email, but it would be quite convenient to have an API to query all revoked versions. Especially for the parties who have not been the members since the very beginning.